Adding CAPX doc (with CAPO example) (#8)
giantswarm/gitops-template: docs/add_cluster_app.md

If a template base for your provider is already here, you may skip this part.

**Bear in mind**, this is not a complete guide of how to create a perfect base, but rather a mere summary of basic steps
needed to move forward. Hence, instructions here will not always be precise in telling you what to change, as this can
strongly depend on resources involved, how much out of them you would like to include into a base, etc.

1. Export provider' and CAPx' names you are about to create, for example `capo` and `openstack`:

    ```sh
    export CAPX=capo
    export PROVIDER=openstack
    ```

1. Create a directory structure:

    ```sh
    mkdir -p bases/clusters/${CAPX}/template
    ```

1. Create cluster App CR template:

    ```sh
    cat <<EOF > bases/clusters/${CAPX}/template/cluster.yaml
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  labels:
    app-operator.giantswarm.io/version: 0.0.0
  name: \${cluster_name}
  namespace: org-\${organization}
spec:
  catalog: giantswarm
  kubeConfig:
    inCluster: true
  name: cluster-${PROVIDER}
  namespace: org-\${organization}
  version: \${cluster_release}
EOF
      ```

> 1. Create a default apps App CR template, note version of the App is set to `0.1.0` by default:

Author: Sure, I don't mind removing default version here, I am not sure about the matching though, I haven't spotted any default apps vs cluster versions matrix suggesting any relations. How do we define this matching?

> 1. Create a default apps App CR template, note version of the App is set to `0.1.0` by default:

Reviewer: My guess is that we will have to care about it ourselves - that's also why I think we need the isolated Cluster Template repo under our management.

    ```sh
    cat <<EOF > bases/clusters/${CAPX}/template/default_apps.yaml
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  labels:
    app-operator.giantswarm.io/version: 0.0.0
  name: \${cluster_name}-default-apps
  namespace: org-\${organization}
spec:
  catalog: giantswarm
  config:
    configMap:
      name: \${cluster_name}-default-apps-config
      namespace: org-\${organization}
  kubeConfig:
    inCluster: true
  name: default-apps-${PROVIDER}
  namespace: org-\${organization}
  version: \${default_apps_release:=0.1.0}
EOF
    ```

> 1. Create default Apps config, note we do not yet put values in a ConfigMap to have YAML syntax highlighting:

Author: We could put it into a ConfigMap in the base already, but then it would be a string, you lose Github' YAML syntax highlighting, you also lose validation of the YAML syntax, what may be beneficial is some cases.

# Add CAPx workload cluster definition (cluster App)

> Follow the below instructions to store CAPx cluster in the repository. The instructions respect the [repository structure](./repo_structure.md).

Reviewer 2: ```suggestion
Follow the below instructions to store CAPx clusters in the repository. The instructions respect the [repository structure](./repo_structure.md).
```

> This doc will show you both how to add a CAPX cluster definition and how to create a cluster based on one of the ready definitions.

Reviewer 2: ```suggestion
This doc will show you both how to add a CAPx cluster definition and how to create a cluster based on one of the ready definitions.
```

> Adding definition can be done on two levels: templates and version specific, see [create template base](#create-shared-template-base-optional)

Reviewer 2: ```suggestion
Adding definitions can be done on two levels: templates and version specific, see [create template base](#create-shared-template-base-optional)
```

If all you want is to create a new CAPX cluster using an existing definition,
skip to the [creating cluster based on existing bases](#creating-cluster-based-on-existing-bases).

**IMPORTANT**, CAPx configuration utilizes the
[App Platform Configuration Levels](https://docs.giantswarm.io/app-platform/app-configuration/#levels),
in the following manner:

- bases provide default configuration via App' `config` field,
- users provide custom configuration via App' `userConfig` field, that is overlaid on top of `config`.

See more about this approach [here](https://github.com/giantswarm/rfc/tree/main/merging-configmaps-gitops).

## Export environment variables

**Note**, Management Cluster codename, Organization name and Workload Cluster name are needed in multiple places
across this instruction, the least error prone way of providing them is by exporting as environment variables:

```sh
export MC_NAME=CODENAME
export ORG_NAME=ORGANIZATION
export WC_NAME=CLUSTER_NAME
```

## Choose bases

In order to avoid code duplication, it is advised to utilize the
[bases and overlays](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/kustomization/#bases-and-overlays)

> of Kustomize in order to configure cluster.

Reviewer 2: ```suggestion
of Kustomize in order to configure the cluster.
```

This repository comes with some built in bases you can choose from, go to the [bases](../bases/clusters)
directory and search for some that meet your needs, then export their paths with:

```sh
export CLUSTER_PATH=CLUSTER_BASE_PATH
```

If desired base is not there, you can add it. Reference the next section to get a general idea of how to do it.

## Create shared template base (optional)

> **IMPORTANT:** template base cannot serve as a standalone base for cluster creation, it is there to only abstract

Reviewer 2: ```suggestion
**IMPORTANT:** template base cannot serve as a standalone base for cluster creation, it is there only to abstract
```

App CRs that are common to all clusters versions, and to provide basic configuration for default apps App.
It is then used as a base to other bases, which provide an overlay with a specific configuration. This is to avoid
code duplication across bases.

If a template base for your provider is already here, you may skip this part.

**Bear in mind**, this is not a complete guide of how to create a perfect base, but rather a mere summary of basic
steps needed to move forward. Hence, instructions here will not always be precise in telling you what to change,

> as this canstrongly depend on resources involved, how much out of them you would like to include into a base, etc.

Reviewer 2: ```suggestion
as this can strongly depend on resources involved, how much of them you would like to include into a base, etc.
```

    EOF
    ```

1. Create a default apps App CR template, note version of the App is set to `0.1.0` by default:

    ```sh
    cat <<EOF > bases/clusters/${CAPX}/template/default_apps.yaml
    apiVersion: application.giantswarm.io/v1alpha1
    kind: App
    metadata:
      labels:
        app-operator.giantswarm.io/version: 0.0.0
      name: \${cluster_name}-default-apps
      namespace: org-\${organization}
    spec:
      catalog: giantswarm
      config:
        configMap:
          name: \${cluster_name}-default-apps-config
          namespace: org-\${organization}
      kubeConfig:
        inCluster: true
      name: default-apps-${PROVIDER}
      namespace: org-\${organization}
      version: \${default_apps_release:=0.1.0}
    EOF
    ```

1. Create default Apps config, note we do not yet put values in a ConfigMap to have YAML syntax highlighting:

    ```sh
    cat <<EOF > bases/clusters/${CAPX}/template/default_apps_config.yaml
    clusterName: \${cluster_name}
    organization: \${organization}
    EOF
    ```

1. Create the `kustomization.yaml`, note usage of
[ConfigMap Generator](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/configGeneration.md)

> for turning config from the previous step into ConfigMap and placing it under the

Reviewer 2: ```suggestion
for turning config from the previous step into a ConfigMap and placing it under the
```

    | \`cluster_release\` | Cluster App version, reference the [Cluster Openstack](https://github.com/giantswarm/cluster-openstack/releases) for more insight on releases |
    | \`default_apps_release\` | Defaults Apps App version, reference the [Default Apps for Openstack](https://github.com/giantswarm/default-apps-openstack/releases) for more insight on releases |
    EOF
    ```

## Create versioned base (optional)

**IMPORTANT**, versioned bases use a shared template base and overlay it with a preferably generic configuration for a
given cluster version. Versioning comes from the fact that `values.yaml` schema may change over multiple releases,
and although minor differences can be handled in the `userConfig` level, it is advised for the bases to follow major
`values.yaml` schema versions to avoid confusion.

For example, both CAPO bases in this repository, the [v0.5.0](../bases/clusters/capo/<=v0.5.0) and
the [v0.6.0](../bases/clusters/capo/>=v0.6.0), are product of the major changes introduced to the `values.yaml` in
the [cluster-openstack v0.6.0 release](https://github.com/giantswarm/cluster-openstack/releases/tag/v0.6.0).

> **IMPORTANT**, despite the below instructions reference `kubectl-gs` for templating configuration, `kubectl-gs`

Reviewer 2: ```suggestion
**IMPORTANT**, despite the below instructions referencing `kubectl-gs` for templating configuration, `kubectl-gs`
```

generates configuration for the most recent schema only. If you configure a base for older versions of cluster app,
it is advised to check what is generated against the version-specific `values.yaml`.

> 1. Export CAPX' name, provider' name and cluster App' version you are about to create, for example `capo` and `v0.8.0`:

Reviewer 2: ```suggestion
1. Export CAPX name, provider name and cluster App version you are about to create, for example `capo`, `openstack` and `v0.8.0`:
```

    ```sh
    cat <<EOF > kustomization.yaml
    apiVersion: kustomize.config.k8s.io/v1beta1
    commonLabels:
      giantswarm.io/managed-by: flux
    kind: Kustomization
    patchesStrategicMerge:
      - patch_userconfig.yaml
    resources:
    - ../../../../../../../${CLUSTER_PATH}
    - cluster_userconfig.yaml
    EOF
    ```

1. (optional) Repeat the same steps if you need to customize default apps App.

1. Leave the `cluster` directory and go to `workload-clusters`:

    ```sh
    # cd management-clusters/${MC_NAME}/organizations/${ORG_NAME}/workload-clusters
    cd ../../
    ```

1. Edit the Kustomization CR for the workload cluster and assign values to the variables from bases, see example below:

    ```yaml
    # cat ${WC_NAME}.yaml
    apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
    kind: Kustomization
    ...
    spec:
      ...
      postBuild:
        substitute:
          cluster_name: "demo0"
          organization: "gitops-demo"
          cluster_release: "0.8.0"
          default_apps_release: "0.2.0"
      ...
    ```

1. Create a Pull Request with the changes you have just done. Once it is merged, Flux should reconcile resources
you have just created.

> ## Recommended next steps

Reviewer 2: Nice 👍
